### PR TITLE
fabric8 v 4.3.0

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Refactoring

### Description
Updated `fabric8` dependency to version [4.3.0](https://github.com/fabric8io/kubernetes-client/releases/tag/v4.3.0)

### Checklist
- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by scholzj** [comment]
Does Fabric8 now have BouncyCastle as dependency?

**Comment by tombentley** [comment]
Not as a proper dependency (if it were then adding it to the pom.xml would be unnecessary). TBH it's not clear what changed in fabric8 to require this.

**Comment by rohanKanojia** [comment]
@scholzj : It was a bug in client  https://github.com/fabric8io/kubernetes-client/issues/1480 which has been fixed. I would suggest you wait for 4.2.1

**Comment by scholzj** [comment]
@rohanKanojia Great. Thanks for your update. We will wait for 4.2.1.

**Comment by blaghed** [comment]
Hi,

Any traction left on this? 4.2.2 is available and somewhat stable.

**Comment by tombentley** [comment]
@stanlyDoge IIRC you had some fun and games with upgrading fabric8 in the past. Once we've released 0.12.0 please could you look at this to see whether the new release works? I'm disinclined to try to do this in time for 0.12.0 at this point.

**Comment by stanlyDoge** [comment]
Unfortunately, the newest version have some changes in asynchronous behaviour and for example, removing the resource and immediately testing whether the resource was removed fails.

**Comment by tombentley** [comment]
I remember now, but it seems that the fabric8 project didn't really consider that a regression, so it seems unlikely now that it will be fixed. So we need to understand to what extent we really depend on the old behaviour, and figure out what we can do about that. Because in the long term it's not an option to not to be able to upgrade.

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java [hunk 1]
```diff
@@ -146,7 +146,7 @@ public static KafkaTopic toTopicResource(Topic topic, Labels labels) {
                     .build();
         }
 
-        KafkaTopic kt = new KafkaTopicBuilder().withApiVersion("v1")
+        KafkaTopic kt = new KafkaTopicBuilder().withApiVersion("kafka.strimzi.io/v1beta1")
```
Comment (tombentley) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java]: Why is this necessary?

Comment (stanlyDoge) [reply topic-operator/src/main/java/io/strimzi/operator/topic/TopicSerialization.java]: I do not know the exact reason, but `fabric8` client needs also an `apiGroup` now.

**Comment by tombentley** [comment]
We'll need to see the Travis and Jenkins tests passing for this PR before it can be merged.

**Comment by stanlyDoge** [comment]
Yeah, for sure. It seems there is some issue with Connect system tests. I have tried to run them locally and they always time out. I need to investigate further.

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests testcase=KafkaST#testMirrorMaker

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests testcase=KafkaST#testKafkaJBODDeleteClaimsTrueFalse

File: systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java [hunk 2]
```diff
@@ -63,9 +64,23 @@ KubeClient client() {
         return customResourcesWithCascading(Kafka.class, KafkaList.class, DoneableKafka.class);
     }
 
+    @SuppressWarnings("checkstyle:Regexp")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java]: What regexp is that?

File: systemtest/src/main/java/io/strimzi/systemtest/Resources.java [hunk 3]
```diff
@@ -669,7 +673,7 @@ DeploymentBuilder defaultCLusterOperator(String namespace, String operationTimeo
         clusterOperator.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(envVars);
 
         return new DeploymentBuilder(clusterOperator)
-                .withApiVersion("apps/v1")
+                .withApiVersion("extensions/v1beta1")
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/Resources.java]: @scholzj In kube v1.11 I think Deployments are available apps/v1. I know that out existing Fabric8 calls are still using extensions. We should change that, right? 

@stanlyDoge Obviously for this PR it's OK to use extensions if it's needed (though I wonder how it worked before).

File: systemtest/src/main/java/io/strimzi/systemtest/Resources.java [hunk 4]
```diff
@@ -574,9 +578,9 @@ private KafkaTopicBuilder defaultTopic(String clusterName, String topicName, int
                 .endSpec();
     }
 
-    DoneableKafkaTopic topic(KafkaTopic topic) {
+    DoneableKafkaTopic topic(KafkaTopic topic, String namespace) {
```
Comment (tombentley) [inline systemtest/src/main/java/io/strimzi/systemtest/Resources.java]: Why was this change necessary? Can't we assume that the given `topic` has the namespace set already, and throw if not?

File: systemtest/src/test/java/io/strimzi/systemtest/StrimziUpgradeST.java [hunk 5]
```diff
@@ -169,13 +169,15 @@ private void copyModifyApply(File root) {
...
 
     private void deleteInstalledYamls(File root) {
-        Arrays.stream(Objects.requireNonNull(root.listFiles())).sorted().forEach(f -> {
-            if (f.getName().matches(".*RoleBinding.*")) {
-                cmdKubeClient().deleteContent(TestUtils.changeRoleBindingSubject(f, NAMESPACE));
-            } else {
-                cmdKubeClient().delete(f);
-            }
-        });
+        if (root != null) {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/StrimziUpgradeST.java]: How come `root` can now be null?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java [hunk 6]
```diff
@@ -154,6 +154,7 @@ public static Kafka createKafkaCluster(String clusterCmNamespace, String cluster
                     .endKafka()
                     .editZookeeper()
                         .withConfig(zooConfigurationJson)
+                        .withMetrics(metricsCm)
```
[the remark this answers is not in the document]
Comment (stanlyDoge) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java]: There was a bug in tests. Value of `metricsEnabled` was set to true even if no metrics were set. With new client, the correct (false) value was set and therefore some tests were failing because they expected different values.

File: systemtest/src/test/java/io/strimzi/systemtest/StrimziUpgradeST.java (hunk 5 above)
Comment (stanlyDoge) [reply systemtest/src/test/java/io/strimzi/systemtest/StrimziUpgradeST.java]: Not sure... but tests failed at NPE time by time.

File: systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java (hunk 2 above)
Comment (stanlyDoge) [reply systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java]: I tried to rebuild without this suppression and everything seems to be ok.

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by tombentley** [comment]
@scholzj please could you review this when you have time. Mostly the changes are trivial, but note the changes I made in the `AbstractNonNamespacedResourceOperator` in order to make `delete()` 'synchronous' (which it always seemed to be with the previous version of fabric8, but with the upgrade the tests started failing). If we end up merging in its current form we should make similar changes to the `AbstractNamespacedResourceOperator` too.

**Comment by scholzj** [comment]
@tombentley I will get to review this later today. But in general ... I do not think it was ever synchronous. And there are several places in the code where we synchronize this (at least for the mamespaced things such as pods etc.). So changing that in the resource operators should address this also on the places where we synchronize the deletion. It seems a bit confusing to fix it here for `AbstractNonNamespacedResourceOperator` and not for `AbstractNamespacedResourceOperator`. We should at least follow up on this very quickly with another PR.

**Comment by tombentley** [comment]
Another aspect about this that is worth pointing out is that the future is completed on the context thread, rather than the fabric8 thread servicing the watcher. This makes me realise that elsewhere we likely are completing futures on worker pool threads. That's not a problem from the pov of threads blocking etc, but it is incorrect from a thread-safety pov, afaics.

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java [hunk 7]
```diff
@@ -118,15 +119,40 @@ public AbstractNonNamespacedResourceOperator(Vertx vertx, C client, String resou
      * Deletes the resource with the given name
      * and completes the given future accordingly
      */
-    protected Future<ReconcileResult<T>> internalDelete(String name) {
+    private Future<ReconcileResult<T>> internalDelete(String name) {
+
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java]: Why the empty lines here and at the end of the method while you have almost no empty lines within the code?

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java [hunk 8]
```diff
@@ -118,15 +119,40 @@ public AbstractNonNamespacedResourceOperator(Vertx vertx, C client, String resou
...
+        Future<ReconcileResult<T>> result = Future.future();
+        Watcher<T> watcher = new Watcher<T>() {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNonNamespacedResourceOperator.java]: I think there should be a timeout here. Otherwise we risk waiting forever if something goes wrong.

**Comment by tombentley** [comment]
I added support for timeout, and also ensuring that the Watch gets closed, which wasn't being done before. On the downside it's a bit more complicated. On the other hand we can reuse this for other places where we watch for a specific condition, so I've factored this into a separate class.

> Please also make sure that the creation of the cluster roles by the CO works fine after the upgrade.

We have `MainIT` which tests the creation of CR, so I think you meant something specific with "after the upgrade", but IDK what you're referring to exactly.

File: operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceSupport.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                        }
+                    },
+                    true,
+                    ar -> {
+                        doneFuture.handle(ar);
+                    });
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceSupport.java]: Should we at least log the exception? Or are we sure that it can be always ignored?

**Comment by tombentley** [comment]
@strimzi-ci run tests

File: systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java [hunk 10]
```diff
@@ -63,9 +64,23 @@ KubeClient client() {
         return customResourcesWithCascading(Kafka.class, KafkaList.class, DoneableKafka.class);
     }
 
+    // @SuppressWarnings("checkstyle:Regexp")
```
Comment (Frawless) [inline systemtest/src/main/java/io/strimzi/systemtest/AbstractResources.java]: Remove this line if it's not needed?
